src/rdf4cpp/rdf/Serialize.hpp

```diff
@@ -0,0 +1,220 @@
+#ifndef RDF4CPP_SERIALIZE_HPP
+#define RDF4CPP_SERIALIZE_HPP
+
+#include <array>
+#include <cstdio>
+#include <cstring>
+#include <ostream>
+#include <string>
+#include <string_view>
+#include <system_error>
+
+namespace rdf4cpp::rdf {
+
+/**
+ * A cursor into a buffer
+ * Describes remaining length (size()) and current write head (data())
+ *
+ * Memory layout is guaranteed to be identical to:
+ * @code
+ * struct Cursor {
+ *      char *data;
+ *      size_t size;
+ * };
+ * @endcode
+ */
+struct Cursor {
+private:
+    char *data_ = nullptr;
+    size_t size_ = 0;
+
+public:
+    [[nodiscard]] constexpr char *data() const noexcept { return data_; }
+    [[nodiscard]] constexpr size_t size() const noexcept { return size_; }
+
+    constexpr void advance(size_t n) noexcept {
+        data_ += n;
+        size_ -= n;
+    }
+
+    constexpr void repoint(char *new_data, size_t new_size) noexcept {
+        data_ = new_data;
+        size_ = new_size;
+    }
+};
+
+template<typename T>
+concept Serializer = requires (T &ser) {
+    typename T::Output;
+    typename T::Buffer;
+
+    { ser.buffer() } -> std::same_as<typename T::Buffer &>;
+    { ser.cursor() } -> std::same_as<Cursor &>;
+    { ser.finalize() } -> std::convertible_to<typename T::Output>;
+    T::flush(&ser.buffer(), ser.cursor());
```

Comment:
Because that makes it way easier to map to C. I don't think you can cast a normal function pointer to a member function pointer

Reply:
Also making it a member function doesn't work in rust because you would need to mutably borrow members of self multiple times, which is illegal even in unsafe code.

(first mutable borrow when passing cursor to function, second mutable borrow when passing self to the voided self param)

Reply:
OK, I am convinced :)